src/Vault.sol

```diff
@@ -0,0 +1,166 @@
+    ///@param debtAmount The amount of debt asset borrowed
+    event Borrow(address indexed debtAsset, address indexed user, uint debtAmount); 
+
+    ///@dev Emitted on repay()
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling repay()
+    ///@param debtAmount The amount of debt asset being repaid
+    event Repay(address indexed debtAsset, address indexed user, uint debtAmount);
+
+    ///@dev Emitted on withdraw()
+    ///@param collateralAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param collateralAmount The amount of collateral asset withdrawn
+    event Withdraw(address indexed collateralAsset, address indexed user, uint collateralAmount);  
+
+    ///@dev Emitted on liquidation()
+    ///@param collateralAsset The address of the debt asset
+    ///@param debtAsset The address of the debt asset
+    ///@param user The address of the user calling withdraw()
+    ///@param debtToCover The amount of debt the liquidator wants to cover
+    ///@param liquidatedCollateralAmount The amount of collateral received by the liquidator
+    event Liquidation(address indexed collateralAsset, address indexed debtAsset, address indexed user, uint debtToCover, uint liquidatedCollateralAmount);
+    
+    ///@dev To set collateralization level of debt engine
+    ///@notice To allow for margin call at defined level, preventing Vault from bearing risk in fast moving markets
+    uint public immutable bufferDenominator; 
+    uint public immutable bufferNumerator;
+
+    ///@dev Returns decimal places of asset prices as dictated by Chainlink Oracle
+    uint public immutable decimals;
+    
+    constructor(address dai_, address weth_, address priceFeedAddress, uint bufferNumerator_, uint bufferDenominator_) {
+        collateral = IERC20(weth_);
+        debt = IERC20(dai_);
+
+        priceFeed = AggregatorV3Interface(priceFeedAddress);
+        decimals = priceFeed.decimals();
+
+        // collateralization level
+        bufferNumerator = bufferNumerator_;
+        bufferDenominator = bufferDenominator_;
+    }
+
+    ///@dev Users deposit collateral asset into Vault
+    ///@param collateralAmount Amount of collateral to deposit
+    function deposit(uint collateralAmount) external {       
+        deposits[msg.sender] += collateralAmount;
+        bool sent = collateral.transferFrom(msg.sender, address(this), collateralAmount);
+        require(sent, "Deposit failed!");  
+        emit Deposit(address(collateral), msg.sender, collateralAmount);
+    }
+    
+    ///@notice Users borrow debt asset calculated based on collateralization level and their deposits 
+    ///@dev See getMaxDebt() for margined colleteralization colculation
+    ///@param debtAmount Amount of debt asset to borrow
+    function borrow(uint debtAmount) external {             
+        uint maxDebt = getMaxDebt(msg.sender);
```

Comment:
You have to functions to check collateralization, `getMaxDebt` and `getCollateralRequired`, which you use only to get a yes/no answer on whether a vault is sufficiently collateralized. If you make this precondition into a postcondition, and check after `debts[msg.sender] += debtAmount;` you can use the same collateralization check as in `withdraw`.

Alternatively, you can make the collateralization check in `withdraw` a post-condition, and use `getMaxDebt` for the check. But you use `getCollateralRequired` in `liquidate`, so I guess that's the good one.

In short, you can remove one of the two.

For color, [this is how MakerDAO does it](https://github.com/makerdao/dss/blob/17187f7d47be2f4c71d218785e1155474bbafe8a/src/vat.sol#L163).

I think Dan Robinson told me of this one when coding Yield v1 :D